src/main/java/pitt/search/semanticvectors/PSI.java

```diff
@@ -323,102 +326,106 @@ private void trainIncrementalPSIVectors(String iterationTag) throws IOException
+							Vector tmp = objectElementalVector.copy();
+							tmp.bind(predicateElementalVector);
+							return tmp;
+						});
+						subjectSemanticVector.superpose(objToAdd, pWeight * (oWeight + predWeight), null);
+						semanticItemVectors.updateVector(subject, subjectSemanticVector);
+
+						Vector subjToAdd = invBindVectorHash.computeIfAbsent(predicate + "_" + subject, key -> {
+							Vector tmp = subjectElementalVector.copy();
+							tmp.bind(predicateElementalVectorInv);
+							return tmp;
+						});
+						objectSemanticVector.superpose(subjToAdd, pWeight * (sWeight + predWeight), null);
+						semanticItemVectors.updateVector(object, objectSemanticVector);
+
+						if (flagConfig.trainingcycles() > 0) //for experiments with generating iterative predicate vectors
+						{
+							Vector predicateSemanticVector = semanticPredicateVectors.getVector(predicate);
+							Vector predicateSemanticVectorInv = semanticPredicateVectors.getVector(predicate + "-INV");
+							//construct permuted editions of subject and object vectors (so binding doesn't commute)
+							Vector permutedSubjectElementalVector = VectorFactory.createZeroVector(flagConfig.vectortype(), flagConfig.dimension());
+							Vector permutedObjectElementalVector = VectorFactory.createZeroVector(flagConfig.vectortype(), flagConfig.dimension());
+							permutedSubjectElementalVector.superpose(subjectElementalVector, 1, predicatePermutation);
+							permutedObjectElementalVector.superpose(objectElementalVector, 1, predicatePermutation);
+							permutedSubjectElementalVector.normalize();
+							permutedObjectElementalVector.normalize();
+
+							Vector predToAdd = subjectElementalVector.copy();
+							predToAdd.bind(permutedObjectElementalVector);
+							predicateSemanticVector.superpose(predToAdd, sWeight * oWeight, null);
+							semanticPredicateVectors.updateVector(predicate, predicateSemanticVector);
+
+							Vector predToAddInv = objectElementalVector.copy();
+							predToAddInv.bind(permutedSubjectElementalVector);
+							predicateSemanticVectorInv.superpose(predToAddInv, oWeight * sWeight, null);
+							semanticPredicateVectors.updateVector(predicate + "-INV", predicateSemanticVectorInv);
+						}
+					} catch (Throwable e) {
+						logger.info(e.getMessage());
+					} finally {
+						secondLock.unlock();
+						firstLock.unlock();
 					}
-				} catch (Throwable e) {
-					logger.info(e.getMessage());
-				} finally {
-					secondLock.unlock();
-					firstLock.unlock();
-				}
 
-				int currCnt = pc.incrementAndGet();
+					int currCnt = pc.incrementAndGet();
 
-				if (currCnt % 100_000 == 0) {
-					logger.info("Processed " + currCnt + " unique predications ...");
-				}
-			});
+					if (currCnt % 100_000 == 0) {
+						logger.info("Processed " + currCnt + " unique predications ...");
+					}
+				});
+			} catch (InterruptedException | RejectedExecutionException e) {
+				// Do nothing. If execution is aborted or interrupted will be handled later
```

Comment:
Don't like the idea of doing nothing on Exception catch, why not at least log info message?